SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/RemotePossessionComponent.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/RemotePossessionComponent.h"
+#include "LoadBalancing/AbstractLBStrategy.h"
+
+#include "Engine/World.h"
+
+DEFINE_LOG_CATEGORY(LogRemotePossessionComponent);
+
+URemotePossessionComponent::URemotePossessionComponent(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+	, PendingDestroy(false)
+{
+	PrimaryComponentTick.bCanEverTick = true;
+	PrimaryComponentTick.bStartWithTickEnabled = true;
+
+#if ENGINE_MINOR_VERSION <= 23
+	bReplicates = true;
+#else
+	SetIsReplicatedByDefault(true);
+#endif
+}
+
+void URemotePossessionComponent::OnAuthorityGained()
+{
+	AController* Controller = Cast<AController>(GetOwner());
+	ensure(Controller);
+	if (Target == nullptr)
+	{
+		UE_LOG(LogRemotePossessionComponent, Error, TEXT("Target is null"));
+		return;
+	}
+	if (!Target->HasAuthority())
+	{
+		UE_LOG(LogRemotePossessionComponent, Verbose, TEXT("Worker is not authoritative over target: %s"), *Target->GetName());
+		return;
+	}
+	else
+	{
+		if (EvaluatePossess())
+		{
+			UE_LOG(LogRemotePossessionComponent, Log, TEXT("Possess(%s)"), *Target->GetName());
+			Controller->Possess(Target);
```

Comment:
Open question : I am wondering if the possession should be done in EvaluatePossess
On the one hand it would be neat to have everything in a single method, to have checks + possession + operations following possession
On the other hand it is possible in the current code with EvaluatePossess + Posession + OnPossess callback. It's just scattered around.

Reply:
We make an independent `EvaluatePossess` is to let customers implement their evaluation before possess. 
When our customers implement their `Possession` they are not allowed to call `Super::Possession` because we put EvalatePossess+Possession in one function.
This is the reason I prefer to make those processes separate.